device/samples/pnp/pnpTemperatureController.js

```diff
@@ -0,0 +1,256 @@
+  let logLine;
+  if (!!(componentName))
+  {
+    logLine = 'Properties have been reported for component:' + componentName;
+  }
+  else{
+    logLine = 'Properties have been reported for root interface';
+  }
+  deviceTwin.properties.reported.update(patch, function (err) {
+    if (err) throw err;
+    console.log(logLine);
+  });
+};
+
+const helperAttachHandlerForDesiredPropertyPatches = (deviceTwin) => {
+
+  deviceTwin.on('properties.desired', (delta) => {
+    if (delta[thermostat1ComponentName] || delta[thermostat2ComponentName] || delta[deviceInfoComponentName]) {
+    const version = delta.$version;
+    const componentName = Object.keys(delta)[0];
+    let logLine;
+    if (!!(componentName))
+    {
+      logLine = 'Received an update for component: ' + componentName + ' with value: ' + JSON.stringify(delta);
+    }
+    else{
+      logLine = 'Received an update for root interface with value: ' + JSON.stringify(delta);
+    }
+    console.log(logLine);
+    const patch = { [componentName]: {} };
+    Object.entries(delta[componentName]).forEach(([propertyName, propertyValue]) => {
+      if (propertyName !== '__t'){
+        const propertyContent = { value: propertyValue };
+        propertyContent.ac = 200;
+        propertyContent.ad = 'Successfully executed patch';
+        propertyContent.av = version;
+        
+        patch[componentName][propertyName] = propertyContent;
+    }});
+    updateComponentReportedProperties(deviceTwin, patch, componentName);
+ }});
+};
+
+const helperAttachExitListener = async (deviceClient) => {
+  const standardInput = process.stdin;
+  standardInput.setEncoding('utf-8');
+  console.log('Please enter q or Q to exit sample.');
+  standardInput.on('data', (data) => {
+    if (data === 'q\n' || data === 'Q\n') {
+      console.log('Clearing intervals and exiting sample.');
+      clearInterval(intervalToken1);
+      clearInterval(intervalToken2);
+      clearInterval(intervalToken3);
+      deviceClient.close();
+      process.exit();
+    } else {
+      console.log('User Input was : ' + data);
+      console.log('Please only enter q or Q to exit sample.');
+    }
+  });
+};
+async function sendTemperatureTelemetry(deviceClient, componentName, index) {
+  console.log('Sending telemetry message %d from component %s ', index, componentName);
+  
+  const data = JSON.stringify({ temperature: 1 + (Math.random() * 90)});
+  const pnpMsg = new Message(data);
+  if (!!(componentName))
+  {
+    pnpMsg.properties.add(messageSubjectProperty, componentName);
+  }
+  pnpMsg.contentType = 'application/json';
+  pnpMsg.contentEncoding = 'utf-8';
+  await deviceClient.sendEvent(pnpMsg);
+}
+
+async function sendDatasizeTelemetry(deviceClient, componentName, index) {
+  console.log('Sending telemetry message %d from root interface', index);
+  const data = JSON.stringify({ workingset: 1 + (Math.random() * 90)});
+  const pnpMsg = new Message(data);
+  if (!!(componentName))
```

Comment:
>if (!!(componentName)) [](start = 2, length = 22)

I'm a bit confused.  Seems as though the code should assert on a component name being defined.  Or  should simply not take a componentName argument.

Reply:
changed into generic method